### PR TITLE
Don't sleep 250 ms when closing connection, if not needed

### PR DESCRIPTION
**Pull request by adrienverge**
In current code, closing connection with the remote servers make the
caller sleep for 250 ms. Is there a reason, that I don't see?

Anyway, I propose to remove it because it can be a problem in many
situations. For example in my case, requests to my API are delayed by
seconds because my code has to open/close several connections.

### CODE REVIEW

**Comment by bmario** [comment]
This code is mitigation for a bug in aiohttp:

> For a ClientSession with SSL, the application must wait a short duration before closing:
https://docs.aiohttp.org/en/stable/client_advanced.html#graceful-shutdown

And the related github issue: https://github.com/aio-libs/aiohttp/issues/1925 

So if I'm not totally misreading the documentation, that sleep has to stay at least for SSL connections. However, I'm open to other suggestions.

**Comment by adrienverge** [comment]
Thanks, I didn't know that aiohttp particularity. Hum, that's not satisfying...

I still think that *adding 250 ms delay to all connections when closing* is a real burden.

I see 2 other solutions:
1. Detect whether connections has TLS/SSL, and sleep only when needed?
2. Use the hack described at https://github.com/aio-libs/aiohttp/issues/1925#issuecomment-592596034?

**Comment by bmario** [comment]
I'd rather chose option 1 then adding code, which objectively should be part of aiohttp. At least that would mimic the documentation.

I'm wondering though, why is that delay an issue for you? Can't you do one of the following?

- Keep the session open and reuse it
- Don't depend on a particular session being finished
- Just ignore it, as it is async code anyways. And hiding latencies is the whole point of async

**Comment by adrienverge** [comment]
Thanks for your feedback @bmario. I've updated the commit to implement solution 1:

### *Don't sleep 250 ms when closing connection, if not needed*

*In current code, closing connection with the remote servers makes the
caller sleep for 250 ms because of a bug in aiohttp:*

> *For a ClientSession with SSL, the application must wait a short
> duration before closing:
> https://docs.aiohttp.org/en/stable/client_advanced.html#graceful-shutdown*

*This delay is only needed when there are TLS/SSL connections. When it's
not the case, I propose to remove this sleep.*

---

I tested this code on both HTTP and HTTPS connections, using:
```diff
@@ -105,6 +105,7 @@ class RemoteServer(object):
         has_ssl_conn = self._http_session.connector and any(
             any(hasattr(handler.transport, '_ssl_protocol') for handler, _ in conn)
             for conn in self._http_session.connector._conns.values())
+        print('has_ssl_conn = %s' % has_ssl_conn)
         await self._http_session.close()
         if has_ssl_conn:
             await asyncio.sleep(0.250)
```

---

> I'm wondering though, why is that delay an issue for you? Can't you do one of the following?

My use case is a server that takes requests from users (they contain a CouchDB cookie) and has to perform calls to CouchDB *on behalf of* these users. Hence, for each API request, my server has to open a new connection to CouchDB, with different credentials. For that reason, my code looks like:
```python
@routes.get('/my/api')
async def process_incoming_request(couchdb_cookie):
    async with aiocouch.CouchDB('localhost:5984', cookie=couchdb_cookie) as couch:
        users = await couch['_users']
        doc = await users['user_1337']
        return {'key': doc['val']}
    # At this point, no reason to keep ClientSession alive, is it?
```

Maybe my approach is wrong, feel free to comment!

Another point is that I have to make some request as `root`, so I could keep at least *that* connection open, and re-use it inside my routes. But I don't see how. For example the following code works for the first requests, but afterwards the connection is closed:
```python
conn = aiocouch.CouchDB('localhost:5984', cookie=couchdb_cookie)

@routes.get('/route1')
async def route1():
    global conn
    with conn as couch:
        await couch['_users']
```
Is there a way to do this, without closing the connection?

File: aiocouch/remote.py [hunk 1]
```diff
@@ -100,8 +100,14 @@ def __init__(self, server, user=None, password=None, cookie=None, **kwargs):
...
     async def close(self):
+        # If ClientSession has TLS/SSL connections, it is needed to wait 250 ms
+        # before closing, see https://github.com/aio-libs/aiohttp/issues/1925.
+        has_ssl_conn = self._http_session.connector and any(
+            any(hasattr(handler.transport, '_ssl_protocol') for handler, _ in conn)
+            for conn in self._http_session.connector._conns.values())
         await self._http_session.close()
-        await asyncio.sleep(0.250)
+        if has_ssl_conn:
+            await asyncio.sleep(0.250)
```
Comment (bmario) [inline aiocouch/remote.py]: According to the aiohttp documentation, you still need a sleep of 0 if SSL isn't used:

> For a ClientSession without SSL, a simple zero-sleep (await asyncio.sleep(0)) will suffice:

**Comment by bmario** [comment]
First of all, thank you for using aiocouch at all and your efforts in improving it.

The answer for your root connection is simple, don't use the with statement. The sole purpose of the with construct is to close the session when the with scope is left. You can have a look at the [metricq-manager](https://github.com/metricq/metricq-manager/blob/master/metricq_manager/manager.py#L86) how that works. Though I just realized, we never properly close the connection there. :(

For the client connections however, that is a different story. I think the solution you came up is totally fine. However, if you don't want to pay the 250ms for each client, you could make a solution similar to the one I proposed for the root connection. Don't use the with statement, but schedule the closing yourself with `call_soon` and after that you can return immidiatly. Have a look at [this line](https://github.com/metricq/metricq-manager/blob/master/metricq_manager/manager.py#L279). This example is for AMQP, but the principle is the same. In fact, while typing this down, I wonder, if I couldn't make that the default behavior for the context manager anyways 🤔

File: aiocouch/remote.py (hunk 1 above)
Comment (adrienverge) [reply aiocouch/remote.py]: Ah! Thanks. Updated:
```diff
@@ -106,8 +106,7 @@ class RemoteServer(object):
             any(hasattr(handler.transport, '_ssl_protocol') for handler, _ in conn)
             for conn in self._http_session.connector._conns.values())
         await self._http_session.close()
-        if has_ssl_conn:
-            await asyncio.sleep(0.250)
+        await asyncio.sleep(0.250 if has_ssl_conn else 0)
 
     @raises(401, "Authentification failed, check provided credentials.")
     async def _check_session(self):
```

**Comment by adrienverge** [comment]
Thanks a lot for your feedback @bmario.

> The answer for your root connection is simple, don't use the with statement. The sole purpose of the with construct is to close the session when the with scope is left. You can have a look at the [metricq-manager](https://github.com/metricq/metricq-manager/blob/master/metricq_manager/manager.py#L86) how that works. Though I just realized, we never properly close the connection there. :(

It's as simple as that?! Great, I'll do this :)
I guess it's "okay" not to `close()` the connection, if your `Manager` lifetime is the same as the process...

> For the client connections however [...] Don't use the with statement, but schedule the closing yourself with `call_soon` and after that you can return immidiatly.

Clever, that certainly works. But it makes the code a bit heavier, so since I don't use SSL, I'll stick to the nice-looking `with` statement (once this PR is merged, if you accept it).

**Comment by bmario** [comment]
I don't understand why the CI is acting up, in my local tests, it works fine ...

**Pull request merged**